### PR TITLE
join code update

### PR DESCRIPTION
**Pull request by rxin**
This is based on #161 by @chenghao-intel .

### CODE REVIEW

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
@chenghao-intel please take a look at this.

**Comment by rxin** [comment]
I've verified on an ec2 cluster that this new code path is much faster than the old one on the master.

File: src/main/scala/shark/execution/CommonJoinOperator.scala [hunk 1]
```diff
@@ -211,49 +274,15 @@ object CommonJoinOperator {
...
-      addIfNew(joinCond.getLeft())
-      addIfNew(joinCond.getRight())
+  // get the evaluated value(boolean) from the table data (the last element in the array)
+  // true means failed in the join filter testing, we may need to skip it
+  @inline final def filterEval[B](data: B): Boolean = {
+    if(data == null) {
+      true      
+    } else {
+      val fields = data.asInstanceOf[Array[AnyRef]]
+	    fields(fields.length - 1).asInstanceOf[org.apache.hadoop.io.BooleanWritable].get
```
Comment (chenghao-intel) [inline src/main/scala/shark/execution/CommonJoinOperator.scala]: Replace the "tab" with spaces

File: src/main/scala/shark/execution/JoinOperator.scala [hunk 2]
```diff
@@ -46,6 +42,7 @@ class JoinOperator extends CommonJoinOperator[JoinDesc] with ReduceSinkTableDesc
   @transient var tagToValueSer: JHashMap[Int, Deserializer] = _
   @transient var keyDeserializer: Deserializer = _
   @transient var keyObjectInspector: StandardStructObjectInspector = _
+//  @transient val tupleIterator = new this.TupleIterator[Array[AnyRef]]()
```
Comment (chenghao-intel) [inline src/main/scala/shark/execution/JoinOperator.scala]: remove this line

File: src/main/scala/shark/execution/MapJoinOperator.scala [hunk 3]
```diff
@@ -53,6 +57,7 @@ class MapJoinOperator extends CommonJoinOperator[MapJoinDesc] {
 
   @transient val metadataKeyTag = -1
   @transient var joinValues: Array[JList[ExprNodeEvaluator]] = _
+//  @transient val tupleIterator = new this.TupleIterator[Array[Object]]()
```
Comment (chenghao-intel) [inline src/main/scala/shark/execution/MapJoinOperator.scala]: remove this line, too

**Comment by chenghao-intel** [comment]
Thanks @rxin for the clean up, it looks good to me, except some small issues that introduced from my commits.

BTW,  it's based on #161, not #221. :)

**Pull request merged**